### ISSUE TITLE
`docker cp` to a mounted volume using the same ID as that used for the original mount

### ISSUE THREAD

**Issue by MohdAhmad** [moby/moby#37314]
**Description

**Steps to reproduce the issue:**
1. Mount volume to a container
2. `docker cp` file to the container

**Describe the results you received:**
Plugin removed the file. Since the same ID was passed for the docker CP call

**Describe the results you expected:**
For a different ID to be provided for the docker cp calls.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**
```
Client:
 Version:       17.12.1-ce
 API version:   1.35
 Go version:    go1.9.4
 Git commit:    7390fc6
 Built: Tue Feb 27 22:15:20 2018
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.1-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   7390fc6
  Built:        Tue Feb 27 22:17:54 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 17.12.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-862.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.669GiB
Name: test1
ID: G5WH:OYVC:MAFB:GVLE:VVWY:RUDC:ILQ5:COK5:7O5M:DLPT:QGOR:YOKO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Description**

**Steps to reproduce the issue:**
1.
2.
3.

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
(paste your output here)
```

**Output of `docker info`:**

```
(paste your output here)
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
could you show the exact steps to reproduce the issue? I'm not sure I understand the problem description

**Comment by MohdAhmad** [comment]
@thaJeztah in order to track mount references, the nutanix volume plugin uses a combination of `Name` and `ID` passed to the plugin from docker as part of `mount` calls. This works fine for containers mounting volumes because the `ID` is unique for each each request. For instance, `ctr1` mounting `v1` will have the Name `v1` and ID `id1` and `ctr2` mounting the same volume will pass `v1` and `id2`.  
But the issue happens when a `docker cp` is issued, as part of the call the following calls are made to the plugin `mount`, `unmount`, `mount` and `unmount`. If we are copying to `ctr1`, all four of these calls use the same `id1` that was passed to the plugin when the volume was initially mounted to the container. Because of this, the first `unmount` as part of `docker cp` will unmount the volume from the host potentially leading to data corruption.

What we'd like to understand is whether using the combination of `Name` and `ID` is a reasonable way to track mount references in which case this is an issue with the docker daemon. Or whether it is not, in which case we'll need to make changes to the plugin.

**Comment by MohdAhmad** [comment]
From the @cpuguy83's first message on https://github.com/moby/moby/issues/34665 `Each Mount/Unmount pair uses a unique ID so plugins can reference count these mounts.` <- this is not the case. The Mount/Unmount being unique is what we expect but it isn't the case.

**Comment by thaJeztah** [comment]
Just to be sure there has not been a change in the code since; do you see the same issue on Docker 18.03.1-ce?

**Comment by MohdAhmad** [comment]
Sure @thaJeztah. I'll try that out now.

**Comment by MohdAhmad** [comment]
@thaJeztah the issue persists with ```[root@test1 rootfs]# docker version
'Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:20:16 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:23:58 2018
  OS/Arch:      linux/amd64
  Experimental: false```

**Comment by thaJeztah** [comment]
Thanks for checking! @ddebroy this was the issue we were discussing on Slack

**Comment by cpuguy83** [comment]
Yes, looks like if the volume is already mounted for a given container (and as such has a cached mount ID), we just use that ID again.
Shouldn't be difficult to fix, but will require touching a several places to pass through an ID.

**Comment by MohdAhmad** [comment]
@cpuguy83 @thaJeztah thanks! Are there plans on getting the fix into an upcoming docker release?

**Comment by thaJeztah** [comment]
@MohdAhmad don't think someone is working on it right now; if you have an understanding of what needs to happen, would you be interested in contributing?

**Comment by MohdAhmad** [comment]
@thaJeztah I haven't worked on docker before but I'm happy to take a stab at this. Is there a slack channel I could join where I might get guidance?
